### PR TITLE
fast_dev_run can be int

### LINKED ISSUE #4228: [feature] Extend fast_dev_run to accept an integer representing num_batches for unit testing
### Motivation

Currently `fast_dev_run` accept only a boolean value.
 
`fast_dev_run` meaning: Runs 1 batch of train, test and val to find any bugs (ie: a sort of unit test).

```
Trainer(fast_dev_run=True)
```

The new feature will extend fast_dev_run to accept a integer representing the number of batches to use for `fast_dev_run` allowing better coverage for unit testing.
```
Trainer(fast_dev_run=2 )
```

Therefore, 
```
if isinstance(fast_dev_run, int):
    num_batches_for_fast_dev_run = fast_dev_run
    fast_dev_run = fast_dev_run > 0
```

### Pitch

`fast_dev_run ` is really useful for debugging and unit testing. However, sometimes, we want to make aggregation over several batches to make sure everything is correct handled.
Setting `fast_dev_run =2` could allow to run 2 batch of train, test and val to find any bugs (ie: better coverage).

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #4228 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases.
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by SkafteNicki — APPROVED** [review]
Just smaller comments

File: pytorch_lightning/trainer/connectors/debugging_connector.py [hunk 1]
```diff
@@ -32,15 +32,26 @@ def on_init_start(
             fast_dev_run
     ):
 
+        if isinstance(fast_dev_run, int) and fast_dev_run < 0:
+            raise MisconfigurationException(
+                f'fast_dev_run={fast_dev_run} is not a'
+                ' valid configuration. It should be >= 0.'
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/debugging_connector.py:38]: should we take `fast_dev_run=0` as False?

File: pytorch_lightning/trainer/connectors/debugging_connector.py [hunk 2]
```diff
@@ -32,15 +32,26 @@ def on_init_start(
             fast_dev_run
     ):
 
+        if isinstance(fast_dev_run, int) and fast_dev_run < 0:
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/debugging_connector.py:35]: I'll add a separate check above to ensure it should be either bool or integer. user can input anything..float, str..

File: pytorch_lightning/trainer/connectors/debugging_connector.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (SkafteNicki) [reply pytorch_lightning/trainer/connectors/debugging_connector.py:38]: I am okay with that. Then you can ignore some of my comments 👍

**Review by s-rog — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/debugging_connector.py [hunk 3]
```diff
@@ -31,16 +31,32 @@ def on_init_start(
...
         self.trainer.fast_dev_run = fast_dev_run
-        if self.trainer.fast_dev_run:
-            limit_train_batches = 1
-            limit_val_batches = 1
-            limit_test_batches = 1
+
+        if fast_dev_run is True:
+            fast_dev_run = 1
+
+        if fast_dev_run:
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/debugging_connector.py:51]: Not sure: If `fast_dev_run` > 1, I think `fast_dev_run ` it should be set to False, so logging works properly.

**Review by tchaton — CHANGES_REQUESTED** [review]
Great PR ! Concern about logging.

File: pytorch_lightning/trainer/connectors/debugging_connector.py (hunk 3 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/debugging_connector.py:51]: @tchaton you mean if `fast_dev_run > 1` it should follow `log_every_n_steps`??

**Comment by Borda** [comment]
just curious and shall be explained in docs how it then differs from overfitting N badges with `limit_train/val/test_batches`

File: pytorch_lightning/trainer/connectors/debugging_connector.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/debugging_connector.py:51]: Yes, if `fast_dev_run > 1`, I would expect him to follow `log_every_n_steps `. Do you agree ?

File: pytorch_lightning/trainer/connectors/debugging_connector.py [hunk 4]
```diff
@@ -31,16 +31,32 @@ def on_init_start(
...
+            )
 
         self.trainer.fast_dev_run = fast_dev_run
-        if self.trainer.fast_dev_run:
-            limit_train_batches = 1
-            limit_val_batches = 1
-            limit_test_batches = 1
+
+        if fast_dev_run is True:
+            fast_dev_run = 1
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/debugging_connector.py:49]: You could avoid a line with :smile: 

```suggestion
        fast_dev_run = int(fast_dev_run)
```

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/debugging_connector.py]: true, where are my python skills :smile:

File: pytorch_lightning/trainer/connectors/debugging_connector.py (hunk 3 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/debugging_connector.py:49]: is it good now?

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 5]
```diff
@@ -613,7 +613,7 @@ def __gather_result_across_time_and_optimizers(self, epoch_output):
 
     def log_train_step_metrics(self, batch_output):
         # when metrics should be logged
-        if self.should_update_logs or self.trainer.fast_dev_run:
+        if self.should_update_logs or (int(self.trainer.fast_dev_run) == 1):
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:616]: Why is this only done when `== 1`?

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:616]: if `fast_dev_run > 1`, it should follow `log_every_n_steps`, else it will log at every step. https://github.com/PyTorchLightning/pytorch-lightning/pull/4629#pullrequestreview-529414432

File: docs/source/trainer.rst [hunk 6]
```diff
@@ -658,6 +651,16 @@ Under the hood the pseudocode looks like this:
...
     trainer = Trainer(fast_dev_run=True)
 
+    # runs 7 train, val, test batches and program ends
+    trainer = Trainer(fast_dev_run=7)
+
+.. note::
+
+    This argument is a bit different from ``limit_train/val/test_batches``. Setting this argument will
+    disable tuner, logger callbacks like ``LearningRateLogger`` and runs for only 1 epoch. This must be
+    used only for debugging purpose. ``limit_train/val/test_batches`` serves their own purpose and won't
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply docs/source/trainer.rst:661]: `limit_train/val/test_batches only limits the number of batches` is it good? or do you have any suggestions?

**Comment by carmocca** [comment]
Sure :+1:

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/debugging_connector.py [hunk 7]
```diff
@@ -31,16 +31,30 @@ def on_init_start(
...
-        if self.trainer.fast_dev_run:
-            limit_train_batches = 1
-            limit_val_batches = 1
-            limit_test_batches = 1
+        fast_dev_run = int(fast_dev_run)
+
+        if fast_dev_run:
+            limit_train_batches = fast_dev_run
+            limit_val_batches = fast_dev_run
+            limit_test_batches = fast_dev_run
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/debugging_connector.py:52]: Hey @rohitgr7, could you set it there instead
```
# When fast_dev_run > 1, we expect model to logs_every_n_steps, etc..
self.trainer.fast_dev_run = `int(self.trainer.fast_dev_run) == 1`
```
So the rest of the code is cleaner

**Review by tchaton — CHANGES_REQUESTED** [review]
Hey @rohitgr7, thanks for your changes.

I am not a huge fan of `(int(self.trainer.fast_dev_run) == 1)`.

When I use `trainer.fast_dev_run`, I expect the logic to be already encapsulated.

T.C

File: pytorch_lightning/trainer/connectors/debugging_connector.py (hunk 7 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/debugging_connector.py:52]: the reason I did it this way is because of this warning by keeping the actual value of `fast_dev_run` passed by user in the trainer.
`rank_zero_warn(f'Skipping batch size scaler since `fast_dev_run={trainer.fast_dev_run}`', UserWarning)`

will update it as you suggested.

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/debugging_connector.py:52]: @tchaton I don't think this is a good idea. if `fast_dev_run > 1`, we can't just set `trainer.fast_dev_run = False`. Isn't this wrong?? let me try something else

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/debugging_connector.py:56]: let me know if new changes are good.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 8]
```diff
@@ -614,7 +614,7 @@ def __gather_result_across_time_and_optimizers(self, epoch_output):
 
     def log_train_step_metrics(self, batch_output):
         # when metrics should be logged
-        if self.should_update_logs or self.trainer.fast_dev_run:
+        if self.should_update_logs or self.trainer.fast_dev_run is True:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:617]:
```suggestion
        if self.should_update_logs or self.trainer.fast_dev_run:
```
see comment above

File: pytorch_lightning/trainer/connectors/debugging_connector.py (hunk 7 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/debugging_connector.py:56]: fast dev run should not do any logging, agreed.

**Review by awaelchli — CHANGES_REQUESTED** [review]

File: pytorch_lightning/trainer/connectors/debugging_connector.py [hunk 9]
```diff
@@ -31,16 +31,34 @@ def on_init_start(
...
         self.trainer.fast_dev_run = fast_dev_run
-        if self.trainer.fast_dev_run:
-            limit_train_batches = 1
-            limit_val_batches = 1
-            limit_test_batches = 1
+        fast_dev_run = int(fast_dev_run)
+
+        # set fast_dev_run=True when it is 1, used while logging
+        if fast_dev_run == 1:
+            self.trainer.fast_dev_run = True
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/debugging_connector.py:51]: The reason I did this to make it follow `log_every_n_steps` when `fast_dev_run >1` as Thomas suggested. If I just do `if self.fast_dev_run` below it will log at every step, so I made it a strict conditional with `if self.fast_dev_run is True` and made the changes here accordingly.

[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/debugging_connector.py:51]: When fast dev run is set, we prefer not to have logging turned on at all. fast dev run is there to quickly debug crashing code in any stage of training, and so polluting the logs with these runs is not desired.

[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/debugging_connector.py:51]: on the master, it does log in case of `fast_dev_run=True`. You want me to change that?? @awaelchli cc: @tchaton

[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/debugging_connector.py:51]: oh, that's unfortunate :( Then we may need to follow up on this in a new PR / issue because I am fairly confident that this is not intended behaviour. The logic to differentiate between True and 1 is a bit unsatisfiying in my view.
Thanks for clarification, I'll remove the "request for changes"

**Review by awaelchli — COMMENTED** [review]
removing my "req. for changes"

EDIT: seems doesn't work like that ...

File: pytorch_lightning/trainer/connectors/debugging_connector.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (s-rog) [reply pytorch_lightning/trainer/connectors/debugging_connector.py:51]: You might be wanting to test elaborate logging calls though (this was a use case for me on a previous project)

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by rohitgr7** [comment]
@williamFalcon this needs your approval.

File: docs/source/trainer.rst [hunk 10]
```diff
@@ -632,17 +632,10 @@ fast_dev_run
 
 |
 
-.. raw:: html
-
-    <video width="50%" max-width="400px" controls
```
Comment (williamFalcon) [inline docs/source/trainer.rst:637]: why is the video gone?

File: pytorch_lightning/trainer/training_loop.py [hunk 11]
```diff
@@ -880,7 +880,7 @@ def build_train_args(self, batch, batch_idx, opt_idx, hiddens):
     def save_loggers_on_train_batch_end(self):
         # when loggers should save to disk
         should_flush_logs = self.trainer.logger_connector.should_flush_logs
-        if should_flush_logs or self.trainer.fast_dev_run:
+        if should_flush_logs or self.trainer.fast_dev_run is True:
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/training_loop.py:883]: why are we changing all these to is True?

**Review by williamFalcon — CHANGES_REQUESTED** [review]

File: docs/source/trainer.rst (hunk 10 above)
Comment (rohitgr7) [reply docs/source/trainer.rst:637]: same video was being added twice https://pytorch-lightning.readthedocs.io/en/latest/trainer.html#fast-dev-run

File: pytorch_lightning/trainer/training_loop.py (hunk 11 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/training_loop.py:883]: if `fast_dev_run > 1` the using just `self.trainer.fast_dev_run` will be True always and will generate logs at every step. Others suggested that it should follow `log_every_n_steps` in such case. Also not sure whether `fast_dev_run` should generate logs, since this is happening on master if set to `True`.

**Comment by rohitgr7** [comment]
@williamFalcon mind review??

File: docs/source/trainer.rst (hunk 10 above)
Comment (rohitgr7) [reply docs/source/trainer.rst]: it's fixed now in another PR.

**Review by williamFalcon — APPROVED** [review]
